### ISSUE TITLE
Bugfix: Fix the Clang compilation of MET version 11.0

### ISSUE THREAD

**Issue by JohnHalleyGotway** [dtcenter/MET#2514]
## Describe the Problem ##
METplus discussion dtcenter/METplus#2127 from @climbfuji notes that while MET version 10.1.1 does compile with Clang on MacOS, MET version 11.0.2 does not. Doing so results in a set of compiler errors and warnings.

This tasks for this issue include:
- [x] Fix the main_v11.0 branch to compile via Clang. And then create an 11.0.3 bugfix release.
- [x] Fix the develop branch to compile via Clang.

The remainder of these changes have been moved to issue MET #2564.
- [ ] Update the MET documentation on the develop branch with information about Clang.
- [ ] Add a new GHA test to the develop branch to compile with Clang for each pull request to ensure that it continues working. Note that this will require adding a new tag to the dtcenter/METbaseimage repository to provide the environment in which MET should be compiled via Clang.
- [ ] Consider also enhancing compile_MET_all.sh to support Clang.

As noted in dtcenter/METplus#2127, test the compilation using Clang version 13.

### Expected Behavior ###
MET version 11.0.2 (and all future versions) should compile with Clang.

### Environment ###
Describe your runtime environment:
*1. Machine: Mac Laptop*
*2. OS: MacOS Monterey (version 12.6.2)*
*3. Software version number(s): MET version 11.0.2 or the main_v11.0 branch*

### Relevant Deadlines ###
*List relevant project deadlines here or state NONE.*

### Funding Source ###
*Define the source of funding and account keys here or state NONE.*

### Assignee ###
- [x] Select **engineer(s)** or **no engineer** required
- [x] Select **scientist(s)** or **no scientist** required: No scientist needed

### Labels ###
- [x] Select **component(s)**
- [x] Select **priority**
- [x] Select **requestor(s)**

### Projects and Milestone ###
- [x] Select **Organization** level **Project** for support of the current coordinated release
- [x] Select **Repository** level **Project** for development toward the next official release or add **alert: NEED PROJECT ASSIGNMENT** label
- [x] Select **Milestone** as the next bugfix version

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [x] [METplus](https://github.com/dtcenter/METplus/issues/new/choose), [MET](https://github.com/dtcenter/MET/issues/new/choose), [METdataio](https://github.com/dtcenter/METdataio/issues/new/choose), [METviewer](https://github.com/dtcenter/METviewer/issues/new/choose), [METexpress](https://github.com/dtcenter/METexpress/issues/new/choose), [METcalcpy](https://github.com/dtcenter/METcalcpy/issues/new/choose), [METplotpy](https://github.com/dtcenter/METplotpy/issues/new/choose)
None needed.

## Bugfix Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [ ] Complete the issue definition above, including the **Time Estimate** and **Funding Source**.
- [ ] Fork this repository or create a branch of **main_\<Version>**.
Branch name: `bugfix_<Issue Number>_main_<Version>_<Description>`
- [ ] Fix the bug and test your changes.
- [ ] Add/update log messages for easier debugging.
- [ ] Add/update unit tests.
- [ ] Add/update documentation.
- [ ] Push local changes to GitHub.
- [ ] Submit a pull request to merge into **main_\<Version>**.
Pull request: `bugfix <Issue Number> main_<Version> <Description>`
- [ ] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)** and **Development** issue
Select: **Organization** level software support **Project** for the current coordinated release
Select: **Milestone** as the next bugfix version
- [ ] Iterate until the reviewer(s) accept and merge your changes.
- [ ] Delete your fork or branch.
- [ ] Complete the steps above to fix the bug on the **develop** branch.
Branch name:  `bugfix_<Issue Number>_develop_<Description>`
Pull request: `bugfix <Issue Number> develop <Description>`
Select: **Reviewer(s)** and **Development** issue
Select: **Repository** level development cycle **Project** for the next official release
Select: **Milestone** as the next official version
- [ ] Close this issue.

**Comment by JohnHalleyGotway** [comment]
Notes on compiling with Clang.

1. Install Clang via MacPorts. Start with existing GNU development environment and then run:
```
sudo port install clang-13
export CXX=`which clang`
export CC=`which clang`
```
2. Compile MET's dependent libraries using Clang.
Get the tar files:
```
wget https://dtcenter.ucar.edu/dfiles/code/METplus/MET/installation/tar_files.tgz
tar -zxf tar_files.tgz
```
And adapt the [compile_MET_all.sh](https://raw.githubusercontent.com/dtcenter/MET/main_v11.0/internal/scripts/installation/compile_MET_all.sh) script.
Success with Clang: GSL, BUFRLIB, ZLIB, LIBPNG, G2CLIB, HDF5, NETCDF-C, NETCDF-CXX
Failure with Clang: Jasper (using GNU instead)
Skipped: HDF, HDFEOS, FREETYPE, CAIRO

**Comment by srherbener** [comment]
@JohnHalleyGotway thanks for creating this issue. I am a UCAR employee working on spack-stack for part of my time. Dom (@climbfuji) is my immediate supervisor. I may have a little bit of time to help with this issue if that is desirable.

I'm trying your instructions above for compiling the library dependencies.

I'm interested, for now, in trying to find a reasonable way to get 11.0.1 to compile successfully on the Mac in our spack-stack build process. Some questions I have are:
- When do you think a bug fix will be available?
- What is the intended C++ language standard for MET?
- Do we need C++ GNU extensions?

I can set the language standard/extensions in our flow if necessary. It's not clear to me what Apple Clang is using by default, but that default may not match what is required and be part of the issue.

Thank you for your help with this!

**Comment by srherbener** [comment]
@JohnHalleyGotway I have gotten the MacOS build working using the latest 11.0.2 version. I made two minor code changes to the MET source code (in the build process) as part of this. I'm interested in getting this addressed in the MET code base, and I want to do this in a manner compatible with your methods and practices. My plan is to create a PR in the MET code base with the changes I used and we can work from there. Again, I'm perfectly fine with finding a different approach if I'm not following your methods/practices with my changes.

Does this sound good with you? Thanks!

**Comment by JohnHalleyGotway** [comment]
Stephen,

I was planning tot ask another crack at this next week, June 12th, after
our next development release version. If you have time to work on this on
the main_v11.0 branch before that please feel free. We’re only using the
C++-11 standard in an attempt to enable it to be compiled on a wide variety
of compiler families and versions.

Thanks
John

**Comment by JohnHalleyGotway** [comment]
Yes, that sounds great! Glad to hear you were able to make progress. Please submit a PR to propose those changes to the main_v11.0 branch. We can use that to create another 11.0 bugfix release. And then I can incorporate the changes into develop for inclusion in future releases.

Thanks
John

**Comment by srherbener** [comment]
I've submitted a PR (#2556) containing the patches I used to get the spack-stack process to work properly. I tried to fill out the PR description as best I could, but please advise and I can make updates as needed. Thanks!

**Comment by climbfuji** [comment]
Thanks @JohnHalleyGotway and @srherbener for fixing the MET 11 build for macOS.

**Comment by JohnHalleyGotway** [comment]
@climbfuji my plan was to also wait for a fix for #2578 from @hsoh-u prior to cutting the 11.0.3 bugfix release.

But I wanted to check in with you about that. Is there a date/deadline on your end driving the need for us to cut the 11.0.3 bugfix release soon?

**Comment by climbfuji** [comment]
No rush. We have the bug fixes that are needed for previous versions 11.0.x in spack-stack, so we can build those.

**Comment by climbfuji** [comment]
But it would be good to know when 11.0.3 is release so that we can add it to spack-stack and not apply the bug fixes from that version onwards.

**Comment by JohnHalleyGotway** [comment]
Reopening this issue since it needs to be fixed again. This was noted by @srherbener in this release acceptance testing [comment](https://github.com/dtcenter/METplus/discussions/2218#discussioncomment-6545195).

The needed changes were merged into develop by dtcenter/MET#2563, but those changes were partially overridden the following day by dtcenter/MET#2570!

Reopening this issue to re-do this change for the main_v11.1 branch. Note that we only need to do this in main_v11.1 since we'll merge all changes from main_v11.1 into develop after the 11.1.0 release.
